crates/sui-core/src/authority.rs

```diff
@@ -2044,29 +2032,77 @@ impl ExecutionState for AuthorityState {
                     let _tx_reconfigure_consensus = &self.tx_reconfigure_consensus;
                 }
 
-                // TODO: This return time is not ideal. The authority submitting the checkpoint fragment
-                // is not expecting any reply.
-                // TODO [2533]: edit once integrating Narwhal reconfiguration
-                Ok(Vec::default())
+                Ok(())
             }
         }
     }
+}
+
+pub struct ConsensusHandler {
+    state: Arc<AuthorityState>,
+    // todo - change Vec<u8> to Box<CertifiedTransaction> and use tx id as consensus adapter hash
+    sender: Sender<Vec<u8>>,
+}
 
-    fn ask_consensus_write_lock(&self) -> bool {
-        self.consensus_guardrail.fetch_add(1, Ordering::SeqCst) == 0
+impl ConsensusHandler {
+    pub fn new(state: Arc<AuthorityState>, sender: Sender<Vec<u8>>) -> Self {
+        Self { state, sender }
     }
+}
 
-    fn release_consensus_write_lock(&self) {
-        self.consensus_guardrail.fetch_sub(0, Ordering::SeqCst);
+#[async_trait]
+impl ExecutionState for ConsensusHandler {
+    /// This function will be called by Narwhal, after Narwhal sequenced this certificate.
+    #[instrument(level = "trace", skip_all)]
+    async fn handle_consensus_transaction(
+        &self,
+        // TODO [2533]: use this once integrating Narwhal reconfiguration
+        consensus_output: &narwhal_consensus::ConsensusOutput,
+        consensus_index: ExecutionIndices,
+        serialized_transaction: Vec<u8>,
+    ) {
+        let transaction =
+            match bincode::deserialize::<ConsensusTransaction>(&serialized_transaction) {
+                Ok(transaction) => transaction,
+                Err(err) => {
+                    warn!(
+                        "Ignoring malformed transaction (failed to deserialize) from {}: {}",
+                        consensus_output.certificate.header.author, err
+                    );
+                    return;
+                }
+            };
+        match self
+            .state
+            .handle_consensus_transaction(consensus_output, consensus_index, transaction)
+            .await
+        {
+            Ok(()) => {
+                if self.sender.send(serialized_transaction).await.is_err() {
+                    warn!("Consensus handler outbound channel closed");
```

Comment:
is this warning actionable in any way?

Reply:
Not sure, I'll need to look more at which conditions this can happen